R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+}
+
+#' Attach the latest available lockbox version to a package
+add_latest_version_in_lockbox <- function(package) {
+  package$latest_version_in_lockbox <- max_package_version(
+    list.files(file.path(lockbox_library(), package)))
+  package
+}
+
+#' Take a vector of versions and find the max without coercing to package_version
+max_package_version <- function(versions) {
+  if (length(versions) == 0) return(NULL)
+  formatted_versions <- package_version(as.character(versions))
+  versions[which(formatted_versions == max(formatted_versions))[1]]
+}
+
+#' Check a dependency list for inclusion in the lockfile and replace the package
+#' with the locked version if it does appear there.  Also throw
+#' an error if we require a dependency version greater than that specified by
+#' the lockfile.
+replace_with_lock <- function(package, lock) {
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- Filter(function(l) l$name == package$name, lock)[[1]]
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
+      package_version(as.character(locked_package$version))) {
+        stop(paste0("Dependency: \'", package$name, ", Version: ", package$version
+          , " is required by package ", package$parent_package
+          , ", but lockbox is locked at version: "
+          , as.character(locked_package$version)
+          , ". Please update your lockfile accordingly"))
+    }
+    package <- locked_package
+  } else {
+    package$is_dependency_package <- TRUE
+  }
+  if (!"remote" %in% names(package) || is.na(package$remote)) {
+    package$remote <- "CRAN"
+  }
+  package <- as.locked_package(package)
+  if (package$is_dependency_package && !package$name %in%
+    as.character(installed.packages(priority = "base")[,1])) {
+      package$latest_version <- package$latest_version_in_lockbox %||%
+        package$latest_version %||% get_latest_version(package)
+  }
+  package
+}
+
+get_latest_version <- function(package) {
+  if (package$remote == "CRAN") {
+    get_available_cran_version(package)
+  } else {
+    version_from_remote(package)
+  }
+}
+
+#' Combine two lists of dependencies via version comparisons.  Keep packages
+#' found in list1 on the left side of the entirety of list2, while moving
+#' the parent package to the space after it's rightmost dependency found in list2.
+combine_dependencies <- function(list1, list2, current_parent) {
+  names <- lapply(list(list1, list2), function(lst) vapply(lst, `[[`, character(1), "name"))
+
+  ## Certain packages are no longer on cran but incorporated into R Core
+  core_pkgs <- as.character(installed.packages(priority = "base")[,1])
+  list1 <- list1[!names[[1]] %in% core_pkgs]
+  names[[1]] <- names[[1]][!names[[1]] %in% core_pkgs]
+  list2 <- list2[!names[[2]] %in% core_pkgs]
+  names[[2]] <- names[[2]][!names[[2]] %in% core_pkgs]
```

Comment:
You can avoid this duplication by working with `lists <- list(list1, list2)` instead.